### ISSUE TITLE
Unpublished repeaters items have 2048 instead of 2049 for status

### ISSUE THREAD

**Issue by adrianbj** [processwire/processwire-issues#36]
### Short description of the issue

As per the title
### Expected behavior

I would expect that unpublished repeater items would also have StatusOn(1) and therefore be 2049.
### Actual behavior

They end up with 2048 
### Optional: Suggestion for a possible fix

Not sure if this needs a fix because I don't really know if it's intended or not. It seems inconsistent to me, but I am guessing there is a valid reason, so maybe this is more of a question that an issue, but I am not sure.

**Comment by ryancramerdesign** [comment]
The two statuses are equivalent. In some cases, I'll use them for temporary runtime purposes but don't recall if I'm doing that here or not. In any case, there's no need to worry about the differing statuses, they should be inconsequential.

**Comment by adrianbj** [comment]
Unfortunately it doesn't seem to be inconsequential in this case - a repeater item with 2049 shows as "New", rather than unpublished. If you are creating an unpublished repeater via the API you need to make sure it is set to 2048 and not 2049 or it simply doesn't work as none of content of the repeater is shown.

**Comment by adrianbj** [comment]
@isellsoap - any chance of re-opening this please. I do think there is more to it than Ryan mentioned. I could be wrong of course, but I would really appreciate a little more clarification on my last point.

Perhaps @ryancramerdesign will still remember to get back to me, but I know how many Github notifications he must get and it must be hard to keep track of them all. Having it open again will be an unavoidable reminder :)
Thanks!

**Comment by ryancramerdesign** [comment]
Taking a closer look, it does look like FieldtypeRepeater uses the Page::statusOn for some identification. How are you creating your Repeater pages? Make sure you are using one of the built-in methods that are setup for this purpose, i.e. 
~~~~~
$item = $page->repeater_field->makeNewItem();
~~~~~
or this
~~~~~
$field = $fields->get('repeater_field');
$item = $field->type->getBlankRepeaterPage($page, $field);
~~~~~
This is good to do not only for making sure it has the right status for repeaters, but also because it will use existing blank ready pages when available. If you are already using one of these and still seeing the wrong status, please let me know.

**Comment by adrianbj** [comment]
Yeah I noticed that use of statusOn. I guess I thought that the inconsistency was still a little weird.

I have put together a one click script that converts a PageTable field to a Repeater or Repeater Matrix field. It converts everything beautifully including all fields, templates, and of course the content of each of the item sub-pages. The thing is that I am not using `makeNewItem` because I am changing the parent of sub-page from the PageTable parent to the relevant Repeater parent (with some config changes), so I think in this case my hack to manually remove StatusOn is probably the best solution.

Anyway, I have a working solution and as long as you are happy, then I guess this isn't really a problem.

**Comment by adrianbj** [comment]
Sorry @ryancramerdesign - more thoughts on this - it currently seems impossible to use either of these on a repeater item, even if they were created via the admin GUI:
```
$item->addStatus(Page::statusUnpublished);
$item->removeStatus(Page::statusUnpublished);
```
Am I missing something?

**Comment by ryancramerdesign** [comment]
Adrian, check out the comments at the top of the FieldtypeRepeater.module file, it has a few notes in there about how Repeater pages use Page statuses and what the different combinations mean. I'm not positive this is the reason for what you are seeing, so you might also want to double check that the repeater items originated from an unformatted value (i.e. outputFormatting == false).

**Comment by isellsoap** [comment]
@adrianbj Does @ryancramerdesign’s suggestion fix this issue for you? Can it be closed?

**Comment by adrianbj** [comment]
I have read the notes about what the different statuses mean, but I still think something is broken when making changes to the status via the API.

Here's a bunch of sequential screenshots that hopefully shows the issue clearly.

**1) Initial state - everything setup completely via the admin GUI:**

![screen shot 2016-11-06 at 10 52 58 am](https://cloud.githubusercontent.com/assets/871211/20040606/3949137e-a40f-11e6-94f4-23b8eadf9e71.png)

Looking at the pages via PHPMyAdmin I see that the status of each is:
`#1` - 1
`#2` - 2048

**2) Now we run this code:**

```
$p = $pages->get(1);
$p->of(false);
foreach($p->repeatertest as $item) $item->removeStatus(Page::statusUnpublished);
$p->save("repeatertest");
```

which results in this:
![screen shot 2016-11-06 at 10 56 08 am](https://cloud.githubusercontent.com/assets/871211/20040637/a908e52c-a40f-11e6-9d9f-b56e028eadb0.png)
which looks fine, but now the status of each page is:
`#1` - 0
`#2` - 1

**3) Now we run this code:**

```
$p = $pages->get(1);
$p->of(false);
foreach($p->repeatertest as $item) $item->addStatus(Page::statusUnpublished);
$p->save("repeatertest");
```

which results in this:
![screen shot 2016-11-06 at 10 58 52 am](https://cloud.githubusercontent.com/assets/871211/20040658/0d25dd08-a410-11e6-849c-5234b965df6d.png)

which seems weird because the first one is now "New" and the status of each page is:
`#1` - 2048
`#2` - 2049

Does that help to explain the problem? Or am I still missing something?

**Comment by CalleRosa40** [comment]
I agree with @adrianbj that this behavior is inconsistent and quite confusing. Running $repeaterItem->addStatus(Page::statusUnpublished) should really unpublish the item. Without having to remove statusOn, that is. It would be great if this were fixed, especially since Repeaters seem to receive an overhaul at the moment anyway.

Apart from that, @ryancramerdesign and all collaborators, ProcessWire is a totally awesome system. Thanks a lot for it!

**Comment by netcarver** [comment]
@adrianbj Hi Adrian, I note this issue is quite old now, is it still relevant or has it been fixed/superseded?

**Comment by adrianbj** [comment]
Hey @netcarver - I think it's a still a mess that needs cleaning up - it's a weird exception which makes API manipulation not work as expected. It certainly hasn't been fixed AFAIK.

**Comment by joer80** [comment]
I can confirm this is still an issue.  Even if you create the repeater item with the admin, to unpublish via api you must do this or it doest work right:

$l->addStatus(Page::statusUnpublished);
$l->removeStatus(Page::statusOn);

**Comment by adrianbj** [comment]
Thanks for your input @joer80 - this has been around for a long time :)

**Comment by ryancramerdesign** [comment]
@joer80 you've got it mostly right but you'll want to set the status directly rather than using addStatus/removeStatus calls. That's because addStatus/removeStatus are what you want to use when you want to add or remove some status without affecting other statuses that might be present. That's not what you want here since repeater combines the unpublished status with others to mean different things (see [here](https://github.com/processwire/processwire/blob/dev/wire/modules/Fieldtype/FieldtypeRepeater/FieldtypeRepeater.module#L22)), which likely don't apply to your API usage. (Though it's good to know what they are, just in case they are useful for your purpose). So to unpublish, keep it simple and just set the status directly:
~~~~
$item->status = Page::statusUnpublished;
~~~~ 
or to publish set status to 0 (no status) or 1 (statusOn flag):
~~~~
$item->status = Page::statusOn;
~~~~

**Comment by ryancramerdesign** [comment]
Closing this one out. As noted in the thread, the correct way to set a repeater item's unpublished status via API is to set the status directly rather than using addStatus/removeStatus:

```php
$item->status = Page::statusUnpublished; // unpublish
$item->status = Page::statusOn; // publish
$pages->save($item);
```

The `addStatus()`/`removeStatus()` approach doesn't work as expected here because repeater items use combined status flags internally. The workaround above is reliable across all versions. If someone wants to propose a specific improvement to make addStatus/removeStatus transparent for repeater items, a new focused issue would be the right place for it.

-Claude

**Comment by adrianbj** [comment]
Proposed the improvement invited above in processwire/processwire#348 — `RepeaterPage` now maintains the `statusOn` flag inside `addStatus()`/`removeStatus()`, so `$item->addStatus(Page::statusUnpublished)` genuinely unpublishes an item (2048, no silent re-publish on the next editor save) and `removeStatus()` restores `statusOn` like the editor does. Ready pages, all internal transitions, and direct `$item->status` assignment are unchanged, with tests covering each case.
